worker/spending.go

```diff
@@ -0,0 +1,100 @@
+package worker
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"go.sia.tech/core/types"
+	"go.sia.tech/renterd/api"
+	"go.sia.tech/renterd/internal/tracing"
+	"go.uber.org/zap"
+)
+
+const keyContractSpendingRecorder contextKey = "ContractSpendingRecorder"
+
+type (
+	ContractSpendingRecorder interface {
+		record(fcid types.FileContractID, cs api.ContractSpending)
+	}
+
+	contractSpendingRecorder struct {
+		bus           Bus
+		flushInterval time.Duration
+		logger        *zap.SugaredLogger
+
+		mu                          sync.Mutex
+		contractSpendings           map[types.FileContractID]api.ContractSpending
+		contractSpendingsFlushTimer *time.Timer
+	}
+)
+
+func RecordContractSpending(ctx context.Context, fcid types.FileContractID, cs api.ContractSpending) {
+	if sr, ok := ctx.Value(keyContractSpendingRecorder).(ContractSpendingRecorder); ok {
+		sr.record(fcid, cs)
+		return
+	}
+	panic("no spending recorder attached to the context") // developer error
+}
+
+func WithContractSpendingRecorder(ctx context.Context, sr ContractSpendingRecorder) context.Context {
+	return context.WithValue(ctx, keyContractSpendingRecorder, sr)
+}
+
+func (w *worker) newContractSpendingRecorder() *contractSpendingRecorder {
+	return &contractSpendingRecorder{
+		bus:               w.bus,
+		contractSpendings: make(map[types.FileContractID]api.ContractSpending),
+		flushInterval:     w.busFlushInterval,
+		logger:            w.logger,
+	}
+}
+
+func (sr *contractSpendingRecorder) record(fcid types.FileContractID, cs api.ContractSpending) {
+	sr.mu.Lock()
+	defer sr.mu.Unlock()
+
+	// Add spending to buffer.
+	sr.contractSpendings[fcid] = sr.contractSpendings[fcid].Add(cs)
+
+	// If a thread was scheduled to flush the buffer we are done.
+	if sr.contractSpendingsFlushTimer != nil {
+		return
+	}
+	// Otherwise we schedule a flush.
+	sr.contractSpendingsFlushTimer = time.AfterFunc(sr.flushInterval, func() {
+		sr.mu.Lock()
```

Comment:
`nit` wouldn't `time.AfterFunc(sr.flushInterval, sr.managedFlush)` be cleaner, I've noticed this before but didn't comment on it at the time but since it's popping up more and more now perhaps it makes sense to update all occurrences? seems like we never want to call `flush` without holding the lock anyway

Reply:
You did comment on it and the reason against it was that one time we only call `flush` under a lock but the other time we do a check and then call `flush`. Both need to happen under the same lock.